src/libtools/examples/merging.cpp

```diff
@@ -0,0 +1,173 @@
+#include <merging/automergestrategy.hpp>
+#include <merging/onesidestrategy.hpp>
+
+
+using namespace kdb;
+using namespace kdb::tools::merging;
+using namespace std;
+
+int main()
+{
+	KeySet ours;
+	KeySet theirs;
+	KeySet base;
+
+	// the root of the subtree containing our keys (i.e. our side of the merge)
+	Key oursRoot ("user/ours", KEY_END);
+
+	// the root of the subtree containing their keys (i.e. their side of the merge)
+	Key theirsRoot ("user/theirs", KEY_END);
+
+	// the root of the subtree containing the base keys (i.e. the common ancestor of the merge)
+	Key baseRoot ("user/base", KEY_END);
+
+	// the root of the subtree that will contain the merge result
+	Key resultRoot ("user/result", KEY_END);
+
+	// Step 1: retrieve clean KeySets containing only those
+	// keys that should be merged. This is a bit trickier than
+	// it seems at first. Have a look at the documentation of kdbGet
+	// for detailed information
+	// things to note:
+	//   * use blocks with local KDB instances so we don't have to worry about
+	//     writing the keys back
+	//   * remove the root key itself from the result KeySet because it usually
+	//     contains the mounted filename and cannot be merged anyway
+	{
+		KDB lkdb;
+		lkdb.get (ours, oursRoot);
+		ours = ours.cut (oursRoot);
+		ours.lookup(oursRoot, KDB_O_POP);
+	}
+	{
+		KDB lkdb;
+		lkdb.get (theirs, theirsRoot);
+		theirs = theirs.cut (theirsRoot);
+		ours.lookup(oursRoot, KDB_O_POP);
+	}
+	{
+		KDB lkdb;
+		lkdb.get (base, baseRoot);
+		base = base.cut (baseRoot);
+		ours.lookup(oursRoot, KDB_O_POP);
+	}
+
+
+	// Step 2: Make sure that no keys reside below the intended merge result root
+	// Usually the merge can be either aborted if this is the case or the existing
+	// keys can be overwritten.
+	KeySet resultKeys;
+	kdb::KDB kdb;
+	kdb.get (resultKeys, resultRoot);
+
+	KeySet discard = resultKeys.cut (resultRoot);
+	if (discard.size () != 0)
+	{
+		// handle existing keys below the result root
+		return -1;
+	}
+
+	ThreeWayMerge merger;
+
+	// Step 3: Decide which resolution strategies to use. The stratgies are registered
+	// with the merge and applied in order as soon as a conflict is detected. If a strategy
+	// marks a conflict as resolved, no further strategies are consulted. Therefore the order
+	// in which they are registered is absolutely crucial. With this chaining the strategies
+	// remain simple, but can be combined to powerful resolution strategies.
+	// Have a look at the strategy documentation for further details on what they do and how they work.
+	// The unit tests also provide insight into how the strategies work.
+
+	// TODO:
+	// Currently the MetaMergeStrategy has to be registered first. Otherwise the meta information
+	// of merged keys is lost. This is due to the way that strategies are currently implemented and
+	// is subject to change.
+
+	// TODO:
+	// Currently the strategies have to be instantiated and freed manually. This is also subject to change
```

Comment:
Hello,

you wrote:
> I just used them in the example for now and reverted to raw pointers for
> passing because I am not sure if I really understand unique_ptr yet.
> However, using them required me to enable C++11 in CMake. Is this really
> intended?

No: we have this unique_ptr/auto_ptr hack that picks the correct one depending 
on C++11 availablility. We should have that hack in the header file, and users 
should use a typedef and do not need to worry.

best regards